java-util/src/main/java/io/druid/java/util/common/granularity/SegmentGranularity.java

```diff
@@ -0,0 +1,450 @@
+        "'dt'=yyyy-MM-dd-HH",
+        "'y'=yyyy/'m'=MM/'d'=dd/'h'=HH",
+        "'y'=yyyy/'m'=MM/'d'=dd/'H'=HH"
+    ),
+    SIX_HOUR(HOUR),
+    DAY(
+        "'dt'=yyyy-MM-dd",
+        "'y'=yyyy/'m'=MM/'d'=dd",
+        "'y'=yyyy/'m'=MM/'d'=dd"
+    ),
+    WEEK(DAY),
+    MONTH(
+        "'dt'=yyyy-MM",
+        "'y'=yyyy/'m'=MM",
+        "'y'=yyyy/'m'=MM"
+    ),
+    YEAR(
+        "'dt'=yyyy",
+        "'y'=yyyy",
+        "'y'=yyyy"
+    );
+
+    private final String hiveFormat;
+    private final String lowerDefaultFormat;
+    private final String defaultFormat;
+
+    GranularityType(
+        final String hiveFormat,
+        final String lowerDefaultFormat, final String defaultFormat
+    )
+    {
+      this.hiveFormat = hiveFormat;
+      this.lowerDefaultFormat = lowerDefaultFormat;
+      this.defaultFormat = defaultFormat;
+    }
+
+    GranularityType(GranularityType granularityType)
+    {
+      this(
+          granularityType.getHiveFormat(),
+          granularityType.getLowerDefaultFormat(),
+          granularityType.getDefaultFormat()
+      );
+    }
+
+    static SegmentGranularity createSegmentGranularity(String str)
+    {
+      return createSegmentGranularity(str, null, null);
+    }
+
+    static SegmentGranularity createSegmentGranularity(String str, DateTime origin, DateTimeZone tz)
+    {
+      GranularityType granularityType = GranularityType.valueOf(str);
+
+      switch (granularityType) {
+        case SECOND: return new PeriodSegmentGranularity(new Period("PT1S"), origin, tz);
+        case MINUTE: return new PeriodSegmentGranularity(new Period("PT1M"), origin, tz);
+        case FIVE_MINUTE: return new PeriodSegmentGranularity(new Period("PT5M"), origin, tz);
+        case TEN_MINUTE: return new PeriodSegmentGranularity(new Period("PT10M"), origin, tz);
+        case FIFTEEN_MINUTE: return new PeriodSegmentGranularity(new Period("PT15M"), origin, tz);
+        case HOUR: return new PeriodSegmentGranularity(new Period("PT1H"), origin, tz);
+        case SIX_HOUR: return new PeriodSegmentGranularity(new Period("PT6H"), origin, tz);
+        case DAY: return new PeriodSegmentGranularity(new Period("P1D"), origin, tz);
+        case WEEK: return new PeriodSegmentGranularity(new Period("P1W"), origin, tz);
+        case MONTH: return new PeriodSegmentGranularity(new Period("P1M"), origin, tz);
+        case YEAR: return new PeriodSegmentGranularity(new Period("P1Y"), origin, tz);
+        default:
+          throw new IAE("[%s] granularity not supported with strings. Try with Period instead", str);
+      }
+    }
+
+    // Note: This is only an estimate based on the values in period.
+    // This will not work for complicated periods that represent say 1 year 1 day
+    static GranularityType estimatedGranularityType(Period period) {
+      int[] vals = period.getValues();
+      BitSet bs = new BitSet();
+      for (int i = 0; i < vals.length; i++) {
+        if (vals[i] != 0) {
+          bs.set(i);
```

Comment:
The BitSet exists to throw an `IAE` for compound periods. B/c compound periods don't have a matching "hive", "default" or "lowerDefault" format.